### PR TITLE
Chore: Adding create/update/delete endpoints for comments

### PR DESCRIPTION
**Pull request by jeremypress**

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @jeremypress has signed the CLA. Thanks for the pull request!

**Comment by jeremypress** [comment]
Tests on the way

File: src/api/Base.js [hunk 1]
```diff
@@ -165,54 +193,152 @@ class Base {
...
         throw new Error('Implement me!');
     }
 
+    /**
+     * Attaches success and error callbacks for a given API call
+     * @param {Function} successCallback - Success function
+     * @param {Function} errorCallback - Error function
+     */
+    attachCallbacks(successCallback: Function, errorCallback: Function) {
+        if (typeof successCallback === 'function') {
```
Comment (DanDeMicco) [inline src/api/Base.js]: Due to the fact that flow checks that the arguments are a function, this helper function isnt necessary

File: src/api/Base.js [hunk 2]
```diff
@@ -165,54 +193,152 @@ class Base {
...
+            this.errorHandler(error);
+        }
+    }
+
+    /**
+     * Generic API put
+     *
+     * @param {string} id - The file id
+     * @param {string} url - The url to post to
+     * @param {Object} data - The data to post
```
Comment (DanDeMicco) [inline src/api/Base.js]: put not post, for url as well

File: src/api/Base.js [hunk 3]
```diff
@@ -165,54 +193,152 @@ class Base {
...
+     * @param {Function} successCallback - The success callback
+     * @param {Function} errorCallback - The error callback
+     * @param {Object} data optional data to delete
+     */
+    async delete(
+        id: string,
+        url: string,
+        successCallback: Function,
+        errorCallback: Function,
+        data?: Object = {}
```
Comment (DanDeMicco) [inline src/api/Base.js]: Maybe see if you can remove the ? since you are defaulting

File: src/api/Comments.js [hunk 4]
```diff
@@ -5,13 +5,15 @@
...
+import { getTypedFileId } from '../util/file';
+import { PERMISSION_CAN_COMMENT, PERMISSION_CAN_EDIT_COMMENT, PERMISSION_CAN_DELETE_COMMENT } from '../constants';
 
 class Comments extends OffsetBasedAPI {
     /**
-     * API URL for comments
+     * API URL for comments on a file
      *
-     * @param {string} [id] - a box file id
+     * @param {string} [id] - A box file id
```
Comment (DanDeMicco) [inline src/api/Comments.js]: Whats with the brackets?

File: src/api/Comments.js [hunk 5]
```diff
@@ -30,7 +32,15 @@ class Comments extends OffsetBasedAPI {
             return;
         }
 
-        const comments = data.entries.map((comment: Comment) => {
+        // There is no response data when deleting a comment
+        if (!data) {
+            return;
```
Comment (DanDeMicco) [inline src/api/Comments.js]: You should call the successCallback

Comment (DanDeMicco) [reply src/api/Comments.js]: Also, change data: any -> data:Object

File: src/api/Comments.js [hunk 6]
```diff
@@ -40,6 +50,143 @@ class Comments extends OffsetBasedAPI {
...
     };
+
+    /**
+     * API URL for comments endpoint
+     *
+     * @param {string} [id] - A box comment file id
+     * @return {string} base url for comments
+     */
+    getCommentsUrl(id?: string): string {
+        const suffix: string = id ? `/${id}` : '';
```
Comment (DanDeMicco) [inline src/api/Comments.js]: Maybe can make it look prettier

const url = ${this.getBaseApiUrl()}/comments
return id ? url/{id} : url

File: src/api/Comments.js [hunk 7]
```diff
@@ -40,6 +50,143 @@ class Comments extends OffsetBasedAPI {
...
+        successCallback,
+        errorCallback
+    }: {
+        file: BoxItem,
+        message: string,
+        taggedMessage?: string,
+        successCallback: Function,
+        errorCallback: Function
+    }): Promise<void> {
+        const { id = '', permissions } = file;
```
Comment (DanDeMicco) [inline src/api/Comments.js]: Why are you defaulting id?

File: src/api/Item.js [hunk 8]
```diff
@@ -155,7 +155,7 @@ class Item extends Base {
      * @param {Boolean} recursive - True for folders
```
Comment (DanDeMicco) [inline src/api/Item.js]: Since you touched this function, remove this since the param isnt there :)

File: src/api/Comments.js (hunk 7 above)
Comment (jeremypress) [reply src/api/Comments.js]: flow doesn't understand that my try catch will guarantee that ID is available by the time we get to line 99. So it thinks i'm possibly calling getTypedFileId on `undefined`. Trying to refactor around this

File: src/api/Comments.js [hunk 9]
```diff
@@ -40,6 +50,143 @@ class Comments extends OffsetBasedAPI {
...
+
+        try {
+            this.checkItem(id, permissions);
+            this.checkPermission(permissions, PERMISSION_CAN_COMMENT);
+        } catch (e) {
+            errorCallback(e);
+            return Promise.reject();
```
Comment (priyajeet) [inline src/api/Comments.js]: Should not return promises unless we are actually trying to read it.

File: src/api/Comments.js [hunk 10]
```diff
@@ -40,6 +50,143 @@ class Comments extends OffsetBasedAPI {
...
+        file,
+        commentId,
+        successCallback,
+        errorCallback
+    }: {
+        file: BoxItem,
+        commentId: string,
+        successCallback: Function,
+        errorCallback: Function
+    }): Promise<void> {
```
Comment (priyajeet) [inline src/api/Comments.js]: no async/await in here, no need to return

Comment (priyajeet) [reply src/api/Comments.js]: same goes for other functions

File: src/api/Base.js [hunk 11]
```diff
@@ -165,54 +193,152 @@ class Base {
...
+    ): Promise<void> {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.attachCallbacks(successCallback, errorCallback);
+
+        // Make the XHR request
+        try {
+            const { responseData } = await this.xhr.post({ id, url, data });
```
Comment (priyajeet) [inline src/api/Base.js]: shouldn't this be { data } since thats what axios sends.
response.data

File: src/api/Base.js [hunk 12]
```diff
@@ -165,54 +193,152 @@ class Base {
...
+        errorCallback: Function,
+        data?: Object = {}
+    ): Promise<void> {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.attachCallbacks(successCallback, errorCallback);
+
+        // Make the XHR request
```
Comment (priyajeet) [inline src/api/Base.js]: It looks like all these non-get functions are same, only thing thats diff is this.xhr[operation]

**Pull request merged**